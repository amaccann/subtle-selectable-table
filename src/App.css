#root {
  max-width: 100%;
  margin: 0 auto;
  padding: 2rem;
  text-align: center;
}

.pending-alert {
  background-color: var(--variant-warning);
  padding: 8px 16px;
  border-radius: 16px;
  color: #fff;
  opacity: 1;
  animation: 500ms 1 normal fade-in;
}

.table-actions {
  margin-bottom: 16px;
}

.table {
  border-spacing: 0;
  margin-top: 16px;
  border: 1px solid var(--table-border-color);
  text-align: left;
  font-size: 14px;
}

.table th, .table td {
  padding: 12px;
  vertical-align: middle;
}

@keyframes fade-in {
  from {
    opacity: 0;
  }

  to {
    opacity: 1;
  }
}

@media screen and (max-width: 920px) {
  #root {
    padding: 0;
  }

  .table {
    font-size: 12px;
  }

  .table-actions {
    justify-content: space-between !important;
  }
}