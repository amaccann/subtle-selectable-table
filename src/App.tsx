import { useEffect, useState } from 'react'
import './App.css'
import { FileData, FileStatus } from './api/data'
import Checkbox from './components/Checkbox';
import TableRow from './components/TableRow';
import Box from './components/Box';
import Button from './components/Button';

function App() {
  const [files, setFiles] = useState<FileData[]>([]);
  const [selected, setSelected] = useState<string[]>([]);

  const areAllSelected: boolean = selected.length === files.length;
  const isIndeterminate: boolean = Boolean(selected.length && selected.length < files.length);
  const pendingAvailable: FileData[] = files.filter((f) => f.status === FileStatus.Scheduled && selected.includes(f.name));
  const noDownloadsAvailable = selected.length === pendingAvailable.length;

  useEffect(() => {
    (async () => {
      /**
       * Just did this to "fake" a render ticket of loading the data
       * Realistically this would be coming from an API, and would have
       * a "loading" state. Had intended to use Netlify Functions to 
       * make a quick little fetch request....
       */
      const {default: data} = await import('./api/data');
      setFiles(data);
    })();
  }, []);

  const onChangeSelectAll = () => {
    if (!areAllSelected) {
      setSelected(files.map(file => file.name));
    } else {
      setSelected([]);
    }
  };

  const onClickDownload = () => {
    const filesToDownload = files.filter(
      ({name, status}) => selected.includes(name) && status === FileStatus.Available,
    );

    const message = filesToDownload
      .reduce(
        (arr, file) => arr + `${file.device}\n  Â· (${file.path}) \n`,
        'Downloading the following:\n'
      );
    alert(message);
  };

  const onToggleFile = (fileName: string) => {
    if (selected.includes(fileName)) {
      setSelected((s) => s.filter((name) => name !== fileName));
    } else {
      setSelected((s) => s.concat(fileName));
    }
  };

  return (
    <>
      <Box className="table-actions" gap={32}>
        <Box gap={8}>
          <Checkbox
            checked={areAllSelected}
            id="select-all"
            indeterminate={isIndeterminate}
            label="Select all"
            onChange={onChangeSelectAll}
          />
          <label title="Selected total" htmlFor="select-all">{!selected.length ? 'None selected' : `Selected ${selected.length}`}</label>
        </Box>
        
        <Button
          disabled={!selected.length || noDownloadsAvailable}
          label="Download selected"
          onClick={onClickDownload}>
          Download selected
        </Button>
      </Box>
      {pendingAvailable.length ? (
        <div className="pending-alert">
          ({pendingAvailable.map(({name}) => name).join(',')}
          <span>&nbsp;not yet available for download</span>)
        </div>
      ) : null}
      <table className="table">
        <thead>
          <tr>
            <th />
            <th>Name</th>
            <th>Device</th>
            <th>Path</th>
            <th>Status</th>
          </tr>
        </thead>
        <tbody>
          {files.map((file: FileData) => {
            return (
              <TableRow
                file={file}
                key={file.name}
                onToggle={onToggleFile}
                selected={selected}
              />

            );
          })}
        </tbody>
      </table>
      
    </>
  )
}

export default App
