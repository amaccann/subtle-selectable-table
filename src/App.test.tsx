import {fireEvent, render, waitFor} from '@testing-library/react'
import App from './App';

describe('Main app portal', () => {
  const renderComponent = async () => {
    const result = await render(<App />);
    await waitFor(() => expect(result.getByText('smss.exe')).toBeInTheDocument());
    return result;
  };

  beforeEach(() => {
    global.alert = jest.fn();
  });

  it('renders app, taking snapshot', async () => {
    const result = await renderComponent();
    expect(result.asFragment()).toMatchSnapshot();
  });

  it('toggles a single [first] row', async () => {
    const {container} = await renderComponent();
    const firstRow = container.querySelector('.table-row input');

    // Select the row
    fireEvent.click(container.querySelector('.table-row input')!);
    expect(firstRow?.getAttribute('aria-checked')).toEqual('true');

    // Unselect the row
    fireEvent.click(container.querySelector('.table-row input')!);
    expect(firstRow?.getAttribute('aria-checked')).toEqual('false');
  });

  it('selects all the Documents', async () => {
    let selectedRows;

    const result = await renderComponent();
    const selectAllBtn = result.container.querySelector('.table-actions input[type="checkbox"]');

    // Select all
    fireEvent.click(selectAllBtn!);
    selectedRows = result.container.querySelectorAll('.table input[aria-checked="true"]');
    expect(selectedRows.length).toEqual(6);

    // De-select all
    fireEvent.click(selectAllBtn!);
    selectedRows = result.container.querySelectorAll('.table input[aria-checked="false"]');
    expect(selectedRows.length).toEqual(6);
  });

  it('downloads all the selected documents', async () => {
    const result = await renderComponent();
    const selectAllBtn = result.container.querySelector('.table-actions input[type="checkbox"]');
    const downloadBtn = result.getByText(/Download selected/i);

    // Select all
    fireEvent.click(selectAllBtn!);
    fireEvent.click(downloadBtn!);

    expect(global.alert).toHaveBeenCalled();
  });
});