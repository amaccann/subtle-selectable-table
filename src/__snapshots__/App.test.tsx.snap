// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main app portal renders app, taking snapshot 1`] = `
<DocumentFragment>
  <div
    class="table-actions"
    style="display: flex; align-items: flex-start; gap: 32px; justify-content: flex-start;"
  >
    <div
      style="display: flex; align-items: flex-start; gap: 8px; justify-content: flex-start;"
    >
      <input
        aria-checked="false"
        aria-label="Select all"
        checked=""
        class="checkbox"
        id="select-all"
        title="Select all"
        type="checkbox"
      />
      <label
        for="select-all"
        title="Selected total"
      >
        None selected
      </label>
    </div>
    <button
      aria-label="Download selected"
      class="button"
      disabled=""
    >
      Download selected
    </button>
  </div>
  <table
    class="table"
  >
    <thead>
      <tr>
        <th />
        <th>
          Name
        </th>
        <th>
          Device
        </th>
        <th>
          Path
        </th>
        <th>
          Status
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="table-row"
        data-can-download="false"
        data-selected="false"
      >
        <td>
          <input
            aria-checked="false"
            aria-label="smss.exe"
            class="checkbox"
            title="smss.exe"
            type="checkbox"
          />
        </td>
        <td>
          smss.exe
        </td>
        <td>
          Mario
        </td>
        <td>
          \\Device\\HarddiskVolume2\\Windows\\System32\\smss.exe
        </td>
        <td>
          <div
            style="display: flex; align-items: center; gap: 0px; justify-content: flex-start;"
          >
            <span
              style="margin-left: 4px;"
            >
              Scheduled
            </span>
          </div>
        </td>
      </tr>
      <tr
        class="table-row"
        data-can-download="true"
        data-selected="false"
      >
        <td>
          <input
            aria-checked="false"
            aria-label="netsh.exe"
            class="checkbox"
            title="netsh.exe"
            type="checkbox"
          />
        </td>
        <td>
          netsh.exe
        </td>
        <td>
          Luigi
        </td>
        <td>
          \\Device\\HarddiskVolume2\\Windows\\System32\\netsh.exe
        </td>
        <td>
          <div
            style="display: flex; align-items: center; gap: 0px; justify-content: flex-start;"
          >
            <span
              class="dot dot--success"
            />
            <span
              style="margin-left: 4px;"
            >
              Available
            </span>
          </div>
        </td>
      </tr>
      <tr
        class="table-row"
        data-can-download="true"
        data-selected="false"
      >
        <td>
          <input
            aria-checked="false"
            aria-label="uxtheme.dll"
            class="checkbox"
            title="uxtheme.dll"
            type="checkbox"
          />
        </td>
        <td>
          uxtheme.dll
        </td>
        <td>
          Peach
        </td>
        <td>
          \\Device\\HarddiskVolume1\\Windows\\System32\\uxtheme.dll
        </td>
        <td>
          <div
            style="display: flex; align-items: center; gap: 0px; justify-content: flex-start;"
          >
            <span
              class="dot dot--success"
            />
            <span
              style="margin-left: 4px;"
            >
              Available
            </span>
          </div>
        </td>
      </tr>
      <tr
        class="table-row"
        data-can-download="false"
        data-selected="false"
      >
        <td>
          <input
            aria-checked="false"
            aria-label="aries.sys"
            class="checkbox"
            title="aries.sys"
            type="checkbox"
          />
        </td>
        <td>
          aries.sys
        </td>
        <td>
          Daisy
        </td>
        <td>
          \\Device\\HarddiskVolume1\\Windows\\System32\\aries.sys
        </td>
        <td>
          <div
            style="display: flex; align-items: center; gap: 0px; justify-content: flex-start;"
          >
            <span
              style="margin-left: 4px;"
            >
              Scheduled
            </span>
          </div>
        </td>
      </tr>
      <tr
        class="table-row"
        data-can-download="false"
        data-selected="false"
      >
        <td>
          <input
            aria-checked="false"
            aria-label="cryptbase.dll"
            class="checkbox"
            title="cryptbase.dll"
            type="checkbox"
          />
        </td>
        <td>
          cryptbase.dll
        </td>
        <td>
          Yoshi
        </td>
        <td>
          \\Device\\HarddiskVolume1\\Windows\\System32\\cryptbase.dll
        </td>
        <td>
          <div
            style="display: flex; align-items: center; gap: 0px; justify-content: flex-start;"
          >
            <span
              style="margin-left: 4px;"
            >
              Scheduled
            </span>
          </div>
        </td>
      </tr>
      <tr
        class="table-row"
        data-can-download="false"
        data-selected="false"
      >
        <td>
          <input
            aria-checked="false"
            aria-label="7za.exe"
            class="checkbox"
            title="7za.exe"
            type="checkbox"
          />
        </td>
        <td>
          7za.exe
        </td>
        <td>
          Toad
        </td>
        <td>
          \\Device\\HarddiskVolume1\\temp\\7za.exe
        </td>
        <td>
          <div
            style="display: flex; align-items: center; gap: 0px; justify-content: flex-start;"
          >
            <span
              style="margin-left: 4px;"
            >
              Scheduled
            </span>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
